build/src/lib.rs

```
@@ -148,52 +286,59 @@ pub struct BinaryBuilder {
+            } else {
+                out
+            })
+        };
+
+        let tokenizer_out = path_transform(self.out_dir.join(tokenizer_filename(lang_code)))?;
+        let rules_out = path_transform(self.out_dir.join(rules_filename(lang_code)))?;
 
         let mut did_not_find_binaries = false;
 
-        for (binary, out) in [
+        for (binary, out) in &[
             (Binary::Tokenizer, &tokenizer_out),
             (Binary::Rules, &rules_out),
-        ]
-        .iter()
-        {
-            let response = get_binary(&self.version, lang_code, *binary, self.cache_dir.as_ref());
-
-            (match response {
-                Ok(mut reader) => {
-                    let mut bytes = Vec::new();
-                    reader
-                        .read_to_end(&mut bytes)
-                        .expect("reading binary bytes failed");
-
-                    fs::write(out, bytes).expect("writing binary to file failed");
-
-                    Ok(())
-                }
-                Err(Error::RequestError(error)) => {
-                    if let Some(404) = error.status().map(|x| x.as_u16()) {
-                        // binaries for this version are not distributed, fall back to building them
-                        did_not_find_binaries = true;
-                        break;
-                    }
-
-                    Err(error.into())
+        ] {
+            let out = out.to_owned().to_owned();
+            if let Err(e) = assure_binary_availability(
+                &self.version,
+                lang_code,
+                *binary,
+                self.cache_dir.as_ref(),
+                self.transform_path_fn.as_ref().map(|x| x.as_ref()),
+                self.transform_data_fn.as_ref().map(|x| x.as_ref()),
+                out,
+            ) {
+                if let Error::BinariesNotFound = e {
+                    did_not_find_binaries = true;
+                    break;
                 }
```

This silently ignores any error happening in `assure_binary_availability` besides `Error::BinariesNotFound`. Should be propagated.

---

Good catch!